system/Filters/Filters.php

```diff
@@ -173,55 +174,135 @@ public function run(string $uri, string $position = 'before')
-                    $this->request,
-                    $this->argumentsClass[$className] ?? null
-                );
+            $result = $class->before(
+                $this->request,
+                $this->argumentsClass[$className] ?? null
+            );
 
-                if ($result instanceof RequestInterface) {
-                    $this->request = $result;
+            if ($result instanceof RequestInterface) {
+                $this->request = $result;
 
-                    continue;
-                }
-
-                // If the response object was sent back,
-                // then send it and quit.
-                if ($result instanceof ResponseInterface) {
-                    // short circuit - bypass any other filters
-                    return $result;
-                }
-                // Ignore an empty result
-                if (empty($result)) {
-                    continue;
-                }
+                continue;
+            }
 
+            // If the response object was sent back,
+            // then send it and quit.
+            if ($result instanceof ResponseInterface) {
+                // short circuit - bypass any other filters
                 return $result;
             }
 
-            if ($position === 'after') {
-                $result = $class->after(
-                    $this->request,
-                    $this->response,
-                    $this->argumentsClass[$className] ?? null
-                );
+            // Ignore an empty result
+            if (empty($result)) {
+                continue;
+            }
 
-                if ($result instanceof ResponseInterface) {
-                    $this->response = $result;
+            return $result;
+        }
 
-                    continue;
-                }
+        return $this->request;
+    }
+
+    private function runAfter(array $filterClasses): ResponseInterface
+    {
+        foreach ($filterClasses as $className) {
+            $class = new $className();
+
+            if (! $class instanceof FilterInterface) {
+                throw FilterException::forIncorrectInterface(get_class($class));
+            }
+
+            $result = $class->after(
+                $this->request,
+                $this->response,
+                $this->argumentsClass[$className] ?? null
+            );
+
+            if ($result instanceof ResponseInterface) {
+                $this->response = $result;
+
+                continue;
+            }
+        }
+
+        return $this->response;
+    }
+
+    /**
+     * Runs required filters for the specified position.
+     *
+     * @return RequestInterface|ResponseInterface|string|null
+     *
+     * @phpstan-param 'before'|'after' $position
+     *
+     * @throws FilterException
+     */
+    public function runRequired(string $position = 'before')
+    {
+        if (! isset($this->config->required[$position]) || $this->config->required[$position] === []) {
```

Comment:
Would it make sense to create **system/Config/Filters.php** with our default set of `$required` to make sure projects have these even if they don't update **app/Config/Filters.php**?

Reply:
So I did.